{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Blast Command Line Tutorial\n",
    "\n",
    "We will discuss basic usage of the blast+ tools (available in graham and cedar or at the [NCBI FTP](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/)).\n",
    "\n",
    "## Objectives / learning outcomes:\n",
    "At the end of this tutorial you shoud be able to:\n",
    "1. Download databases from NCBI \n",
    "2. Create your own custom local database\n",
    "3. Do a basic/intermediate nucleotide blast search\n",
    "4. Change the search parameters to fulfill your needs\n",
    "5. Search for help regarding the available parameters\n",
    "\n",
    "## Prerequisites:\n",
    "Before we start, make sure you went over your bash notes, since we will be using several of the commands we saw in prevoious tutorials. Despite I will touch briefly on the web-based blast, I am assuming you know the basics on how to make blast searches on the [BLAST website](https://blast.ncbi.nlm.nih.gov/Blast.cgi).\n",
    "\n",
    "\n",
    "I will discuss only the `blastn` (nucleotide blast) for time sake, but will comment on other types of blasts briefly.\n",
    "\n",
    "## Outline of the tutorial\n",
    "1. [Introduction to BLAST: Types of databases, searches (this is brief, remember to brush up on this)](#Introduction-to-BLAST)\n",
    "2. [Downloading databases from NCBI FTP with `update_blastdb.pl`](#Downloading-databases-from-NCBI-FTP-with-update_blastdb.pl)\n",
    "3. [Basic `blastn` search](#Basic-blastn-search)\n",
    "4. [Creating local databases with `makeblastdb`](#Creating-local-databases-with-makeblastdb)\n",
    "5. Tunning parameters\n",
    "6. Basic bash manipulation of output\n",
    "\n",
    "## Before we start\n",
    "Log in into your Compute Canada account, create a folder for this tutorial (e.g Blast_tutorial), and open an interactive shell with salloc as we explained last tutorial with 2GB of memory. in the terminal type (remember to change the account for your own group if you are not in Cristescu lab):\n",
    "\n",
    "```bash\n",
    "salloc -A def-mcristes --mem=2000 --time=0-5:0:00\n",
    "module load nixpkgs/16.09  gcc/5.4.0 blast+/2.6.0\n",
    "update_blastdb.pl patnt --decompress\n",
    "```\n",
    "\n",
    "\n",
    "Let it run in the background, we will get back at this, but it takes some time (about 10 mins after allocation)!!\n",
    "\n",
    "## Introduction to BLAST\n",
    "BLAST or Basic Local Alignment Search Tool, is an alignment service available at [NCBI](https://blast.ncbi.nlm.nih.gov/Blast.cgi). As its name exlpains, this software aligns a query sequence (your sequence) with the database sequences, and returns the closest match to your query. It does this in 4 steps (images from [Alarfaj et al.](https://www.doi.org/10.4172/jcsb.1000260)):\n",
    "\n",
    "1. Preprocessing\n",
    "<img src=\"https://raw.githubusercontent.com/jshleap/CristescuLab_misc/master/Tutorials/Blast/img/pre_processing.png\" alt=\"alt text\" width=\"3000\">\n",
    "2. Seeding\n",
    "![alt text](https://raw.githubusercontent.com/jshleap/CristescuLab_misc/master/Tutorials/Blast/img/Seeding-step.png)\n",
    "3. Extension\n",
    "![alt text](https://raw.githubusercontent.com/jshleap/CristescuLab_misc/master/Tutorials/Blast/img/extension_step.png)\n",
    "4. Evaluation\n",
    "![alt text](https://raw.githubusercontent.com/jshleap/CristescuLab_misc/master/Tutorials/Blast/img/evaluation_step.png)\n",
    "\n",
    "Let's check their [webpage](https://blast.ncbi.nlm.nih.gov/Blast.cgi) and run an example blast\n",
    "\n",
    "\n",
    "## Downloading databases from NCBI FTP with update_blastdb.pl\n",
    "\n",
    "Now let's go back to the commands we typed before, by now it should be done! Let's examine the contents of the folder, you should have multiple files with extensions `.nhd`, `.nhi`, `.nhr`, `.nnd`, `.nni`,  `.nog`, `.nsd`, `.nsi`, and `.nsq`. You dont need to worry about their contents, but just in case:\n",
    "- nhr: deflines\n",
    "- nin: indices\n",
    "- nsq: sequence data\n",
    "- nnd: GI data\n",
    "- nni: GI indices\n",
    "- nsd: non-GI data\n",
    "- nsi: non-GI indices\n",
    "\n",
    "If you downloaded the sequences with the update_blastdb.pl script, you will also find a `.nal` file, which is an alias for the database to be searched, as well as two `taxdb.*` binary files with the relationships with taxonomy. However, for us to be able to search taxonomy related information we must download the taxonoy database as well. As you see we have downloaded the `patnt` database which is \"Patent nucleotide sequences. Both patent databases\n",
    "are directly from the USPTO, or from the EPO/JPO via EMBL/DDBJ\". I chose this database more for convenience, since it has more than one olume, but is small enough that we can do the tutorial with. But let's explore what other databases we can use. For that, go to the [NCBI FTP](ftp://ftp.ncbi.nlm.nih.gov/blast/db/). We can see all available volumes and databases to download, but we cannot tell what they are. Let's peak at the README file... BINGO! all the information relating these databases is sumarized in this file. Now, I mentioned that we cannot access taxonomy information without downloading the `taxdb` datatabase, so let's give it a try:\n",
    "```\n",
    "update_blastdb.pl taxdb --decompress\n",
    "```\n",
    "This command should run fast since it is a smaller database. Now we have set up one database to search against. Now we only need our query file in fasta format. For easyness, I've uploaded a working example [here](https://raw.githubusercontent.com/jshleap/CristescuLab_misc/master/Tutorials/Blast/files/sequence.fasta). Explore this file: go to the NCBI website and find it by its accession number. Check its properties, etc. Copy the sequence and do a nucleotide blast in the NCBI website. In the **Algorithm parameters** tab, select **Max target sequences** to 500. Keep the results page open.\n",
    "\n",
    "Now we are set to go to the basic blast.\n",
    "\n",
    "\n",
    "## Basic `blastn` search\n",
    "\n",
    "`blastn` stands for nucleotide blast. It has two mandatory arguments which are `-query` (the fasta file) and `-db` the database where to look. Our first seach will be as follows:\n",
    "\n",
    "```\n",
    "blastn -db /path/to/database/patnt -query /path/to/query/nameoffasta.fasta > blast.hits\n",
    "```\n",
    "Of course you have to change `/path/to/database/patnt` with the actual path where you downloaded the database and the `/path/to/query/nameoffasta.fasta` with the actual path to your file. \n",
    "\n",
    "This command is exacly identical to:\n",
    "\n",
    "```\n",
    "blastn -db /path/to/database/patnt -query /path/to/query/nameoffasta.fasta -out blast.hits\n",
    "```\n",
    "\n",
    "Let's check the summary of options we have:\n",
    "\n",
    "```\n",
    "blastn -h\n",
    "```\n",
    "\n",
    "Some of them are not very informative, let's see the expanded version of the help.\n",
    "\n",
    "```\n",
    "blastn -help\n",
    "```\n",
    "Now we have a very verbose help, where we can see that `-out` allow us to redirect the output to a file. This will become handy later on when we want to manipulate files without writing all intermediate files.\n",
    "\n",
    "The format of the output file is parwise (check the help for *formatting options*). You can change the formatting by passing numbers to the `-outfmt` option. Looking at the help we see that the options are:\n",
    "\n",
    "     0 = Pairwise,\n",
    "     1 = Query-anchored showing identities,\n",
    "     2 = Query-anchored no identities,\n",
    "     3 = Flat query-anchored showing identities,\n",
    "     4 = Flat query-anchored no identities,\n",
    "     5 = BLAST XML,\n",
    "     6 = Tabular,\n",
    "     7 = Tabular with comment lines,\n",
    "     8 = Seqalign (Text ASN.1),\n",
    "     9 = Seqalign (Binary ASN.1),\n",
    "    10 = Comma-separated values,\n",
    "    11 = BLAST archive (ASN.1),\n",
    "    12 = Seqalign (JSON),\n",
    "    13 = Multiple-file BLAST JSON,\n",
    "    14 = Multiple-file BLAST XML2,\n",
    "    15 = Single-file BLAST JSON,\n",
    "    16 = Single-file BLAST XML2,\n",
    "    17 = Sequence Alignment/Map (SAM),\n",
    "    18 = Organism Report\n",
    "\n",
    "The default (Pairwise) shows a list of hits followed by the aligments. This is very similar to what you saw in the webpage. All query-anchored formats (codes 1-4) show the multiple alignment, in slightly different ways. The XML format (code 5) is a special hierarchichal format, similar to webpages. The tabular formats are (in my opinion) the most useful for bioinformaticians. These are tab-delimited files, which are easy to explore with `head`, `tail`, `more`, `less` and `cat` commands, and also are easy to parse with multiple commands and languages. We will work later on with these. Formats 8 and 9 or Seqalign format are (from the docs) \"a collection of segments representing one complete alignment\", and are used for more detailed objectives, where the aligment is the result we are looking for. The format 10 or comma-separated is other very useful format since we can see it in many difefrent programs. Formats 11 - 17 are special formats that fulfill advanced usages of blast, but we arent going into their details since you probably wont use them (I havent!). The last one, *Organism Report* (format 18) produces an interesting table per species and all the sequence hits from that species. This can be very useful to explore taxonomy.\n",
    "\n",
    "Now before we explore a bit more of this formats we will learn how to create our own custom database.\n",
    "\n",
    "\n",
    "## Creating local databases with `makeblastdb`\n",
    "\n",
    "Let's imagine hat you have a set of sequences of interest. This set can be sequenced by you or be a subset of the NCBI databases. There are many reasons why you would like to create your custom database. For example, you only want to search fish, or only prokariotes, etc.. Also, smaller database search is orders of magnitude faster that in the full database. For this exercise, go back to the web blast you made earlier. Select all seqences and download it as a fasta file. Transfer this file to the Graham. For those who dont know how to do this, we can use a command (mac and linux only) called `rsync`. If you are in a windows machine and you have MobaXterm, follow their instructions.\n",
    "\n",
    "```\n",
    "rsync -av seqdump.txt <your-username-graham>@graham.computecanada.ca:/path/where/to/move/it\n",
    "```\n",
    "\n",
    "**Change <your-username-graham> for your actual username, and /path/where/to/move/it, with the actual path to the tutorial folder**\n",
    "\n",
    "Once on the cluster take a look at the file and explore it. Tell me how many lines did you find. Remember that we ask for 500? why the difference?\n",
    "\n",
    "Now let's see what options makeblastdn has:\n",
    "\n",
    "```\n",
    "makeblastdb -help\n",
    "```\n",
    "\n",
    "Now lets create a basic database:\n",
    "\n",
    "```\n",
    "makeblastdb -in seqdump.txt -dbtype nucl -parse_seqids -title Cystic_fibrosis -out cystic_fib.db\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
